File: scripts/update-dbmovie-data.js
import axios from 'axios';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

// --- 环境初始化 ---
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const TMDB_API_KEY = process.env.TMDB_API_KEY;
const DATA_DIR = path.join(process.cwd(), 'data');
const OUTPUT_FILE = path.join(DATA_DIR, 'dbmovie-data.json');

const GENRE_MAP = {
    28: "动作", 12: "冒险", 16: "动画", 35: "喜剧", 80: "犯罪", 99: "纪录片", 18: "剧情", 10751: "家庭", 14: "奇幻", 36: "历史", 27: "恐怖", 10402: "音乐", 9648: "悬疑", 10749: "爱情", 878: "科幻", 10770: "电视电影", 53: "惊悚", 10752: "战争", 37: "西部", 10759: "动作冒险", 10762: "儿童", 10763: "新闻", 10764: "真人秀", 10765: "科幻奇幻", 10766: "肥皂剧", 10767: "脱口秀", 10768: "战争政治"
};

const REGIONS = [
    { title: "全部", limit: 300 },
    { title: "华语", limit: 150 },
    { title: "欧美", limit: 150 },
    { title: "韩国", limit: 150 },
    { title: "日本", limit: 150 }
];

const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

// --- 核心逻辑 ---

async function fetchWithRetry(title, maxRetries = 3) {
    for (let i = 0; i < maxRetries; i++) {
        try {
            const res = await axios.get(`https://api.themoviedb.org/3/search/movie`, {
                params: { api_key: TMDB_API_KEY, query: title, language: 'zh-CN' },
                timeout: 10000
            });
            const matched = res.data.results?.[0];
            if (!matched) return null;

            return {
                id: matched.id.toString(),
                type: "tmdb",
                title: matched.title,
                description: matched.overview,
                posterPath: matched.poster_path ? `https://image.tmdb.org/t/p/w500${matched.poster_path}` : null,
                backdropPath: matched.backdrop_path ? `https://image.tmdb.org/t/p/w500${matched.backdrop_path}` : null,
                rating: matched.vote_average,
                releaseDate: matched.release_date,
                genreTitle: (matched.genre_ids || []).map(id => GENRE_MAP[id]).filter(Boolean).join(', ')
            };
        } catch (err) {
            let waitTime = Math.pow(2, i + 1);
            if (err.response?.status === 429) {
                const retryAfter = parseInt(err.response.headers['retry-after']);
                waitTime = retryAfter ? retryAfter + 1 : waitTime;
            }
            if (i === maxRetries - 1) return null;
            await sleep(waitTime * 1000);
        }
    }
}

async function main() {
    console.log(`\x1b[35m[START]\x1b[0m 开始同步豆瓣电影数据 (ESM 模式)`);
    if (!TMDB_API_KEY) throw new Error("TMDB_API_KEY is missing");

    await fs.mkdir(DATA_DIR, { recursive: true });

    let finalData = {};
    for (const region of REGIONS) {
        console.log(`\n\x1b[36m▶ 正在处理: ${region.title}\x1b[0m`);
        try {
            const res = await axios.get(`https://m.douban.com/rexxar/api/v2/subject/recent_hot/movie`, {
                params: { start: 0, limit: region.limit, type: region.title, score_range: "6,10" },
                headers: {
                    "User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15",
                    "Referer": "https://movie.douban.com/"
                }
            });

            const items = res.data.items || [];
            const results = [];
            // 使用串行+微延迟保护 API
            for (let i = 0; i < items.length; i++) {
                const detail = await fetchWithRetry(items[i].title);
                if (detail) results.push(detail);
                if ((i + 1) % 50 === 0) console.log(`   进度: ${i + 1}/${items.length}`);
                await sleep(150);
            }
            finalData[region.title] = results;
            console.log(`\x1b[32m✅ ${region.title} 完成，匹配成功: ${results.length}\x1b[0m`);
        } catch (e) {
            console.error(`\x1b[31m❌ ${region.title} 失败: ${e.message}\x1b[0m`);
        }
        await sleep(2000);
    }

    await fs.writeFile(OUTPUT_FILE, JSON.stringify(finalData, null, 2));
    console.log(`\n\x1b[32m🎉 数据已写入: ${OUTPUT_FILE}\x1b[0m`);
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});

File: scripts/update-dbmovie-data.py
import os
import asyncio
import aiohttp
import json
import time
from datetime import datetime

# --- 配置 ---
TMDB_API_KEY = os.environ.get('TMDB_API_KEY')
DATA_DIR = os.path.join(os.getcwd(), 'data')
OUTPUT_FILE = os.path.join(DATA_DIR, 'dbmovie-data.json')
CONCURRENCY_LIMIT = 5  # 每秒最大请求数限制

# 缓存已匹配过的电影，避免跨区域重复查询
tmdb_cache = {}

class RateLimiter:
    """简单的令牌桶限流器"""
    def __init__(self, rate):
        self.rate = rate
        self.tokens = rate
        self.updated_at = time.monotonic()

    async def wait(self):
        while self.tokens < 1:
            now = time.monotonic()
            self.tokens += (now - self.updated_at) * self.rate
            self.updated_at = now
            if self.tokens < 1:
                await asyncio.sleep(0.1)
        self.tokens -= 1

limiter = RateLimiter(CONCURRENCY_LIMIT)

async def fetch_tmdb_detail(session, title):
    """异步获取 TMDB 详情，带缓存和重试逻辑"""
    if title in tmdb_cache:
        return tmdb_cache[title]

    await limiter.wait() # 频率控制
    url = "https://api.themoviedb.org/3/search/movie"
    params = {"api_key": TMDB_API_KEY, "query": title, "language": "zh-CN"}
    
    try:
        async with session.get(url, params=params, timeout=10) as resp:
            if resp.status == 429:
                retry_after = int(resp.headers.get("Retry-After", 1))
                await asyncio.sleep(retry_after)
                return await fetch_tmdb_detail(session, title)
            
            if resp.status == 200:
                data = await resp.json()
                results = data.get("results", [])
                if results:
                    matched = results[0]
                    res = {
                        "id": str(matched.get("id")),
                        "title": matched.get("title"),
                        "rating": matched.get("vote_average"),
                        "releaseDate": matched.get("release_date"),
                        "posterPath": f"https://image.tmdb.org/t/p/w500{matched.get('poster_path')}" if matched.get('poster_path') else None,
                    }
                    tmdb_cache[title] = res
                    return res
    except Exception as e:
        pass
    return None

async def process_region(session, region):
    """处理单个区域"""
    print(f"\033[36m▶ 正在处理区域: {region['title']}\033[0m")
    douban_url = "https://m.douban.com/rexxar/api/v2/subject/recent_hot/movie"
    headers = {"User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X)", "Referer": "https://m.douban.com/movie/"}
    params = {"start": 0, "limit": region['limit'], "type": "" if region['title'] == "全部" else region['title']}

    try:
        async with session.get(douban_url, params=params, headers=headers) as resp:
            items = (await resp.json()).get("items", [])
            # 并发执行详情获取
            tasks = [fetch_tmdb_detail(session, item['title']) for item in items]
            results = await asyncio.gather(*tasks)
            matched = [r for r in results if r]
            print(f"   {region['title']} 完成: 匹配 {len(matched)}/{len(items)}")
            return region['title'], matched
    except Exception as e:
        print(f"   [Error] {region['title']} 失败: {e}")
        return region['title'], []

async def main():
    if not os.path.exists(DATA_DIR): os.makedirs(DATA_DIR)
    
    async with aiohttp.ClientSession() as session:
        regions = [
            {"title": "全部", "limit": 200},
            {"title": "华语", "limit": 100},
            {"title": "欧美", "limit": 100}
        ]
        
        final_data = {}
        for region in regions:
            name, data = await process_region(session, region)
            final_data[name] = data
            await asyncio.sleep(1) # 区域间稍微停顿

        with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
            json.dump(final_data, f, ensure_ascii=False, indent=2)

if __name__ == "__main__":
    asyncio.run(main())
